pyquil/api/quantum_computer.py

```diff
@@ -0,0 +1,187 @@
+            necessarily correspond to qubit indices; rather they are the second argument to
+            any MEASURE instructions you've added to your program
+        :param trials: The number of times to run the program.
+        :param symmetrize_readout: Whether to apply readout error symmetrization. If not specified,
+            the class attribute ``symmetrize_readout`` will be used. See
+            :py:func:`run_symmetrized_readout` for a complete description.
+        :return: A numpy array of shape (trials, len(classical_addresses)) that contains 0s and 1s
+        """
+        if symmetrize_readout is None:
+            symmetrize_readout = self.symmetrize_readout
+
+        if not classical_addresses:
+            classical_addresses = get_classical_addresses_from_program(program)
+
+        if symmetrize_readout:
+            return self.run_symmetrized_readout(program, classical_addresses, trials)
+
+        return self.qam.run(program, classical_addresses, trials)
+
+    def run_async(self, program, classical_addresses, trials, symmetrize_readout=None) -> str:
+        """
+        Queue a quil program for running, but return immediately with a job id.
+
+        Use :py:func:`QuantumComputer.wait_for_job` to get the actual job results, probably
+        after queueing up a whole batch of jobs.
+
+        See :py:func:`run` for this function's parameter descriptions.
+
+        :returns: a job id
+        """
+        if symmetrize_readout is None:
+            symmetrize_readout = self.symmetrize_readout
+
+        if not classical_addresses:
+            classical_addresses = get_classical_addresses_from_program(program)
+
+        if symmetrize_readout:
+            raise NotImplementedError("Async symmetrized readout isn't supported")
+
+        return self.qam.run_async(program, classical_addresses, trials)
+
+    def run_symmetrized_readout(self, program, classical_addresses, trials):
+        """
+        Run a quil program in such a way that the readout error is made collectively symmetric
+
+        This means the probability of a bitstring ``b`` being mistaken for a bitstring ``c`` is
+        the same as the probability of ``not(b)`` being mistaken for ``not(c)``
+
+        A more general symmetrization would guarantee that the probability of ``b`` being
+        mistaken for ``c`` depends only on which bit of ``c`` are different from ``b``. This
+        would require choosing random subsets of bits to flip.
+
+        In a noisy device, the probability of accurately reading the 0 state might be higher
+        than that of the 1 state. This makes correcting for readout more difficult. This
+        function runs the program normally ``(trials//2)`` times. The other half of the time,
+        it will insert an ``X`` gate prior to any ``MEASURE`` instruction and then flip the
+        measured classical bit back.
+
+        See :py:func:`run` for this function's parameter descriptions.
```

Comment:
I think it's slightly confusing to have two public methods`run(..., symmetrize_readout)` and `run_symmetrized_readout` with documentation shared between them (and one calling the other). As a user I would prefer a single public API call for this and a single docstring. My suggestion is to make this method private and include the documentation with either `run` or at the class toplevel.

Reply:
I don't see how this is confusing. Why shouldn't one call the other?

Reply:
Well, if I read the docs for `run` it points me to the docs for `run_symmetrized_readout`, which again refers me to the docs for `run`. So to fully understand either function you need to look at both docstrings in parallel.

Reply:
I personally find that sort of cross-referenced documentation easier to grok than having one big wall of text that describes every possible way of running on the device

bear in mind that we'll probably be adding additional `run` variants in the future

Reply:
Ok, works for me